test/hotspot/gtest/nmt/test_vmatree.cpp

```diff
@@ -284,6 +285,138 @@ TEST_VM_F(NMTVMATreeTest, LowLevel) {
+    tree.set_tag(500, 100, mtClassShared);
+    expect_equivalent_form(expected, tree);
+  }
+
+  { // Now let's add in some committed data
+    testrange expected[]{
+        {  0, 100,          mtGC, si, State::Reserved},
+        {100, 225,          mtGC, si, State::Committed},
+        {225, 500,          mtGC, si, State::Reserved},
+        {500, 550, mtClassShared, si, State::Reserved},
+        {550, 560, mtClassShared, si, State::Committed},
+        {560, 565, mtClassShared, si, State::Reserved},
+        {565, 575, mtClassShared, si, State::Committed},
+        {575, 600, mtClassShared, si, State::Reserved}
+    };
+    VMATree tree;
+
+    tree.reserve_mapping(0, 600, rd);
+    // The committed areas
+    tree.commit_mapping(100, 125, rd);
+    tree.commit_mapping(550, 10, rd);
+    tree.commit_mapping(565, 10, rd);
+    // OK, set tag
+    tree.set_tag(0, 500, mtGC);
+    tree.set_tag(500, 100, mtClassShared);
+    expect_equivalent_form(expected, tree);
+  }
+
+  { // Setting the tag for adjacent regions with same stacks should merge the regions
+    testrange expected[]{
+        {0, 200, mtGC, si, State::Reserved}
+    };
+    VMATree tree;
+    Tree::RegionData gc(si, mtGC);
+    Tree::RegionData compiler(si, mtCompiler);
+    tree.reserve_mapping(0, 100, gc);
+    tree.reserve_mapping(100, 100, compiler);
+    tree.set_tag(0, 200, mtGC);
+    expect_equivalent_form(expected, tree);
+  }
+
+  { // Setting the tag for adjacent regions with different stacks should NOT merge the regions
+    NCS::StackIndex si1 = 1;
+    NCS::StackIndex si2 = 2;
+    testrange expected[]{
+        {  0, 100, mtGC, si1, State::Reserved},
+        {100, 200, mtGC, si2, State::Reserved}
+    };
+    VMATree tree;
+    Tree::RegionData gc(si1, mtGC);
+    Tree::RegionData compiler(si2, mtCompiler);
+    tree.reserve_mapping(0, 100, gc);
+    tree.reserve_mapping(100, 100, compiler);
+    tree.set_tag(0, 200, mtGC);
+    expect_equivalent_form(expected, tree);
+  }
+
+  { // Setting the tag in the middle of a range causes a split
+    testrange expected[]{
+        {  0, 100, mtCompiler, si, State::Reserved},
+        {100, 150,       mtGC, si, State::Reserved},
+        {150, 200, mtCompiler, si, State::Reserved}
+    };
+    VMATree tree;
+    Tree::RegionData compiler(si, mtCompiler);
+    tree.reserve_mapping(0, 200, compiler);
+    tree.set_tag(100, 50, mtGC);
+    expect_equivalent_form(expected, tree);
+  }
+
+  { // Setting the tag in between two ranges causes a split
+    testrange expected[]{
+        {  0,  75,       mtGC, si, State::Reserved},
+        { 75, 125,    mtClass, si, State::Reserved},
+        {125, 200, mtCompiler, si, State::Reserved},
+    };
+    VMATree tree;
+    Tree::RegionData gc(si, mtGC);
+    Tree::RegionData compiler(si, mtCompiler);
+    tree.reserve_mapping(0, 100, gc);
+    tree.reserve_mapping(100, 100, compiler);
+    tree.set_tag(75, 50, mtClass);
+    expect_equivalent_form(expected, tree);
+  }
+}
+
```

Comment:
Do we allow address holes? If yes, please add a test where we set_tag over an address hole. If the expected behavior is to ignore the hole, this should be easily tested.

Otherwise, if we assert in case of set_tag over a hole, its fine. We could write a death test to test the assert, but I don't think that is worth the test cycles (death test are expensive).